tpu-client-next/src/client_builder.rs

```diff
@@ -0,0 +1,340 @@
+    reporter_handle: Option<CancellableHandle<()>>,
+}
+
+/// [`ClientBuilder`] is a builder structure to create [`TransactionSender`] along with [`Client`].
+pub struct ClientBuilder {
+    runtime_handle: Option<runtime::Handle>,
+    leader_updater: Box<dyn LeaderUpdater>,
+    bind_target: Option<BindTarget>,
+    identity: Option<StakeIdentity>,
+    num_connections: usize,
+    leader_send_fanout: usize,
+    skip_check_transaction_age: bool,
+    sender_capacity: usize,
+    worker_channel_size: usize,
+    max_reconnect_attempts: usize,
+    report_fn: Option<ReportFn>,
+    cancel_scheduler: CancellationToken,
+    cancel_reporter: CancellationToken,
+}
+
+impl ClientBuilder {
+    pub fn with_leader_updater(leader_updater: Box<dyn LeaderUpdater>) -> Self {
+        Self {
+            runtime_handle: None,
+            leader_updater,
+            bind_target: None,
+            identity: None,
+            num_connections: 64,
+            leader_send_fanout: 2,
+            skip_check_transaction_age: true,
+            worker_channel_size: 2,
+            sender_capacity: 64,
+            max_reconnect_attempts: 2,
+            report_fn: None,
+            cancel_scheduler: CancellationToken::new(),
+            cancel_reporter: CancellationToken::new(),
+        }
+    }
+
+    /// Set the runtime handle for the client. If not set, the current runtime will be used.
+    ///
+    /// Note that if the runtime handle is not set, the caller must ensure that the `build` is
+    /// called in tokio runtime context. Otherwise, `build` will panic.
+    pub fn runtime_handle(mut self, handle: runtime::Handle) -> Self {
+        self.runtime_handle = Some(handle);
+        self
+    }
+
+    /// Set the UDP socket to bind to.
+    pub fn bind_socket(mut self, bind_socket: UdpSocket) -> Self {
+        self.bind_target = Some(BindTarget::Socket(bind_socket));
+        self
+    }
+
+    /// Set the leader send fanout.
+    pub fn leader_send_fanout(mut self, fanout: usize) -> Self {
+        self.leader_send_fanout = fanout;
+        self
+    }
+
+    /// Set the identity keypair for the client.
+    pub fn identity<'a>(mut self, identity: impl Into<Option<&'a Keypair>>) -> Self {
+        self.identity = identity.into().map(StakeIdentity::new);
+        self
+    }
+
+    /// Set the maximum number of cached connections.
+    pub fn max_cache_size(mut self, num_connections: usize) -> Self {
+        self.num_connections = num_connections;
+        self
+    }
+
+    /// Set the cancellation token for the client.
+    ///
+    /// This token is used to create child tokens for the scheduler and reporter tasks. It is useful
+    /// if user wants to immediately cancel all internal tasks, otherwise calling `Client::shutdown`
+    /// is prefered way because it ensures orderly shutdown of internal tasks, see
+    /// `Client::shutdown` for details.
+    pub fn cancel_token(mut self, cancel: CancellationToken) -> Self {
```

Comment:
If this does not ensure orderly shutdown, why would one use this method?

Reply:
To stop all the tasks immediately if needed. But this is a bit hypothetical -- like if a user want to stop all the tasks abruptly and not interested in reporter to give him all the metrics precisely. Does it make sense?

Reply:
it does but probably unnecessary. More API = more mess to maintain. up to you if you want to axe.